docs: Add events API (#59)
CESARBR/knot-babeltower: docs/events.md

</details>

<details>
  <summary>Payload</summary>

  JSON in the following format:

  - `id` **String** thing's ID
  - `name` **String** thing's name

  Example:

  ```json
  {
    "id": "fbe64efa6c7f717e",
    "name": "KNoT Thing"
  }
  ```
</details>

<details>
  <summary>AMQP Binding</summary>

  - AMQP:
    - Exchange:
      - Type: direct
      - Name: device
      - Durable: `true`
      - Auto-delete: `false`
    - Routing key: device.register

</details>

### **device.unregister** <a name="device-unregister"></a>

Event-command to remove a thing from the things registry. The operation response is sent through [`device.unregistered`](#device-registered) event.

<details>
  <summary>Headers</summary>

  - `token` **String** user's token

</details>

<details>
  <summary>Payload</summary>

  JSON in the following format:

  - `id` **String** thing's ID

  Example:

  ```json
  {
    "id": "fbe64efa6c7f717e"
  }
  ```
</details>

<details>
  <summary>AMQP Binding</summary>

  - Exchange:
    - Type: direct
    - Name: device
    - Durable: `true`
    - Auto-delete: `false`
  - Routing key: device.unregister

</details>

### **device.list** <a name="device-list"></a>

> Event-command to list the registered things. It follows the request/reply pattern. After obtaining the things, `babeltower` will send a reply message by using the `correlation_id` property, which is received in the request header, as its `routing_key`. Because of that, considering the **requestor** has created and sent this `correlation_id` in the request, it can also subscribe to receive events that arrive with a `routing_key` equivalent to the `correlation_id`. Therefore, the reply is received by the application that has sent the request, in a **one-to-one** manner.

Reviewer: Awesome documentation :heart: :confetti_ball: 

A quick **suggestion**, tho! `its` is quite ambiguous as it may refer to `babeltower` or `reply message` :grimacing: 

```suggestion
Event-command to list the registered things. It follows the request/reply pattern. After obtaining the things, `babeltower` will send a reply message by using the `correlation_id` property, which was received in the request header, as reply message's  `routing_key`. Because of that, considering the **requestor** has created and sent this `correlation_id` in the request, it can also subscribe to receive events that arrive with a `routing_key` equivalent to the `correlation_id`. Therefore, the reply is received by the application that has sent the request, in a **one-to-one** manner.
```

Reviewer 2: Done.

# Babeltower Events API - v2.0.0

This document describes the events `babeltower` is able to receive and send. They are gruped based on the external clients point of view, i.e. publishing or subscribing to the topics. In each section, it is provided information about the header, payload and protocol binding details of the event.

## Content

* Publish (external clients can publish to):
  - [device.register](#device-register)
  - [device.unregister](#device-unregister)
  - [device.schema.sent](#device-schema-sent)
  - [data.sent](#data-sent)
  - [data.request](#data-request)
  - [data.update](#data-update)
  - [device.list](#device-list)
  - [device.auth](#device-auth)

* Subscribe (external clients can subscribe to):
  - [device.registered](#device-registered)
  - [device.unregistered](#device-unregistered)
  - [device.schema.updated](#device-schema-updated)
  - [data.published](#data-published)
  - [device.[id].data.request](#device-<id>-data-request)

>   - [device.[id].data.update](#device-<id>-data-update)

Reviewer: Very organized :heart_eyes: 

It's a good opportunity to reference the section in question (publish and subscribe)! Also, I suggest using `-` insteado of `*` to avoid [MD004](https://github.com/DavidAnson/markdownlint/blob/v0.19.0/doc/Rules.md#md004) in unordered lists throughout the documentation (no just here).

**Bro tip**: install `prettier` extension and do `ctrl + a`, then `ctrl+shift+i` :smile: 

```suggestion
- [Publish](##Publish) (external clients can publish to):
  - [device.register](#device-register)
  - [device.unregister](#device-unregister)
  - [device.schema.sent](#device-schema-sent)
  - [data.sent](#data-sent)
  - [data.request](#data-request)
  - [data.update](#data-update)
  - [device.list](#device-list)
  - [device.auth](#device-auth)

- [Subscribe](##Subscribe) (external clients can subscribe to):
  - [device.registered](#device-registered)
  - [device.unregistered](#device-unregistered)
  - [device.schema.updated](#device-schema-updated)
  - [data.published](#data-published)
  - [device.[id].data.request](#device-<id>-data-request)
  - [device.[id].data.update](#device-<id>-data-update)
```

Reviewer 2: Nice @lcbm, I think we can follow this style from now on.

Reviewer 2: Done.

-----------------------------------------------------------------

## Publish

This section describes the events that this service can receive from the external applications.

### **device.register** <a name="device-register"></a>

Event-command to register a new thing on the things registry. The operation response is sent through [`device.registered`](#device-registered) event.

<details>
  <summary>Headers</summary>

  * `token` **String** user's token

</details>

<details>
  <summary>Payload</summary>

  JSON in the following format:

  * `id` **String** thing's ID
  * `name` **String** thing's name

  Example:

  ```json
  {
    "id": "fbe64efa6c7f717e",
    "name": "KNoT Thing"
  }
  ```
</details>

<details>
  <summary>Binding</summary>

  * AMQP:
    * Exchange:
      * Type: direct
      * Name: device
      * Durable: `true`
      * Auto-delete: `false`

>       * vhost:

Reviewer: Does `vhost` hold any value? :eyes:

Reviewer 2: Nops.
Done.

>     * Routing key: device.register

Reviewer: Considering we only have AMQP bindings, what do you think about this :point_down:  

```suggestion
  <summary>AMQP Binding</summary>

  - Exchange:
    - Type: direct
    - Name: device
    - Durable: `true`
    - Auto-delete: `false`
    - vhost:
  - Routing key: device.register
```

Reviewer 2: Accepted. Done.

</details>

### **device.unregister** <a name="device-unregister"></a>

Event-command to remove a thing from the things registry. The operation response is sent through [`device.unregistered`](#device-registered) event.

<details>
  <summary>Headers</summary>

  * `token` **String** user's token

</details>

<details>
  <summary>Payload</summary>

  JSON in the following format:

  * `id` **String** thing's ID

  Example:

  ```json
  {
    "id": "fbe64efa6c7f717e"
  }
  ```
</details>

<details>
  <summary>Binding</summary>

  * AMQP:
    * Exchange:
      * Type: direct
      * Name: device
      * Durable: `true`
      * Auto-delete: `false`
      * vhost:
    * Routing key: device.unregister

</details>

### **device.list** <a name="device-list"></a>

> Event-command to list the registered things. It follows the request/reply pattern. After receiving this event, `babeltower` will send a response specifically to the requestor by using the `correlation_id` property, which is receive in the header, as the `routing key`.

Reviewer: Very comprehensive :grimacing: 

Just a minor spellcheck. 
Also, from the reading, I assumed `routing key` was sent in the header. Only from the example that I understood what you actually meant: correlation_id is sent in the header and represents the routing key. Considering both the spell check and the point above, heres a suggestion :point_down: 

```suggestion
Event-command to list the registered things. It follows the request/reply pattern. After receiving this event, `babeltower` will send a response specifically to the requestor through the `correlation_id` value (sent in the header) which is used as the `routing key`.
```

Reviewer 2: In this case, I agree with you the text could be improved but the `correlation_id` *value* is actually used to indicate the reply `routing_key`, i.e. it won't be sent as a header. Let me know if you got it.

I will try to improve this paragraph, thanks.

Reviewer 2: @lcbm take a look, please.

Reviewer: The `correlation_id` **value** is sent in the header, is that right? If this is the case, consider the updated suggestions (just a suggestion, tho... if you have a better options feel free to use that instead).

<details>
  <summary>Headers</summary>

  * `token` **String** user's token
  * `correlation_id` **String** ID to correlate request/reply
  * `reply_to` **String** queue that will process the reply

</details>

<details>
  <summary>Payload</summary>

  JSON in the following format:

  * Empty object

  Example:

  ```json
  {}
  ```

</details>

<details>
  <summary>Binding</summary>

  * AMQP:
    * Exchange:
      * Type: direct
      * Name: device
      * Durable: `true`
      * Auto-delete: `false`
      * vhost:
    * Routing key: `correlation_id`

</details>

### **device.auth** <a name="device-auth"></a>

> Event-command to verify if a thing is authenticated based on its credentials. It follows the request/reply pattern. After receiving this event, `babeltower` will send a response specifically to the requestor by using the `correlation_id` property, which is receive in the header, as the `routing key`.

Reviewer: Same as the previous suggestion :grimacing:

  JSON in the following format:

  * `id` **String** device's ID
  * `token` **String** device's token

  Example:

  ```json
  {
    "id": "fbe64efa6c7f717e",
    "token": "0c20c12e2ac058d0513d81dc58e33b2f9ff8c83d"
  }
  ```
</details>

<details>
  <summary>Binding</summary>

  * AMQP:
    * Exchange:
      * Type: direct
      * Name: device
      * Durable: `true`
      * Auto-delete: `false`
      * vhost:
    * Routing key: `correlation_id`

</details>

### **device.schema.sent** <a name="device-schema-sent"></a>

Event that represents a device sending its schema to the services that are interested. After receiving this event, `babeltower` updates the thing's schema on the registry and send a [`device.schema.updated`](#device-schema-updated) event.

<details>
  <summary>Headers</summary>

  * `token` **String** user's token

</details>

<details>
  <summary>Payload</summary>

  JSON in the following format:

  * `id` **String** thing's ID
  * `schema` **Array** schema items, each one formed by:
    * `sensorId` **Number** sensor ID
    * `valueType` **Number** semantic value type (voltage, current, temperature, etc)
    * `unit` **Number** sensor unit (V, A, W, W, etc)

>     * `typeId` **Number** data value type (boolean, integer, etc)

Reviewer: Could we reference the source where all available `typeId`s, `valueType`s and `unit`s are listed? :stuck_out_tongue:  

Other than code, heres a possible [reference](https://knot-devel.cesar.org.br/doc/thing/unit-type-value.html?highlight=typeid) (gotta make sure its updated, tho).

Reviewer 2: Done.

Event that represents a thing was registered.

<details>
  <summary>Payload</summary>

  JSON in the following format:

  * `id` **String** thing's ID
  * `token` **String** thing's token
  * `error` **String** described the occurred error

  Success example:

  ```json
  {
    "id": "fbe64efa6c7f717e",
    "token": "5b67ce6bef21701331152d6297e1bd2b22f91787",
    "error": null
  }
  ```

  Error example:

  ```json
  {
    "id": "3aa21010cda96fe9",
    "token": "",
    "error": "Device already exists"
  }
  ```

</details>

<details>
  <summary>Binding</summary>

  * AMQP:
    * Exchange:
      * Type: direct
      * Name: device
      * Durable: `true`
      * Auto-delete: `false`
      * vhost:
    * Routing key: device.registered

</details>

### **device.unregistered** <a name="device-unregistered"></a>

Event that represents a thing was removed.

<details>
  <summary>Payload</summary>

  JSON in the following format:

  * `id` **String** thing's ID
  * `error` **String** described the occurred error

  Success example:

  ```json
  {
    "id": "fbe64efa6c7f717e",
    "error": null
  }
  ```

  Error example:

  ```json
  {
    "id": "3aa21010cda96fe9",

>     "error": "Forbidden",

Reviewer: Our errors are being declared in lowercase :stuck_out_tongue: 

```suggestion
    "error": "forbidden",
```

Reviewer 2: Done.

  }
  ```
</details>

<details>
  <summary>Binding</summary>

  * AMQP:
    * Exchange:
      * Type: direct
      * Name: device
      * Durable: `true`
      * Auto-delete: `false`
      * vhost:
    * Routing key: device.unregistered

</details>

### **device.schema.updated** <a name="device-schema-updated"></a>

Event that represents a thing's schema was updated.

<details>
  <summary>Payload</summary>

  JSON in the following format:

  * `id` **String** thing's ID
  * `error` **String** a string with detailed error message

  Success example:

  ```json
  {
    "id": "fbe64efa6c7f717e",
    "error": null
  }
  ```

  Error example:

  ```json
  {
    "id": "3aa21010cda96fe9",

>     "error": "- \"valueType\" must be larger than or equal to 1\n- \"name\" must be a string\n- \"value\" does not contain 1 required value(s)"

Reviewer: From what I recall, as of today, we only return `invalid schema`, due to the way it's currently being validated.

Reviewer 2: Done.

  * AMQP:
    * Exchange:
      * Type: fanout
      * Name: data.published
      * Durable: `true`
      * Auto-delete: `false`
      * vhost:

</details>

### **device.[id].data.request** <a name="device-<id>-data-request"></a>

Event-command to request a specific thing's sensor data after validated.

<details>
  <summary>Headers</summary>

  * `token` **String** user's token

</details>

<details>
  <summary>Payload</summary>

  JSON in the following format:

  * `id` **String** thing's ID
  * `data` **Array** data items to be published, each one formed by:
    * `sensorId` **Number** sensor ID
    * `value` **Number|Boolean|String** sensor value

  Example:

  ```json
  {
    "id": "fbe64efa6c7f717e",
    "data": [
      {
        "sensorId": 1,
        "value": false
      },
      {
        "sensorId": 2,
        "value": 1000
      }
    ]
  }
  ```
</details>

<details>
  <summary>Binding</summary>

  * AMQP:
    * Exchange:
      * Type: direct
      * Name: device
      * Durable: `true`
      * Auto-delete: `false`
      * vhost:
    * Routing Key: `device.<id>.data.request`

</details>

### **device.[id].data.update** <a name="device-<id>-data-update"></a>

> Event-command to update a specific thing's sensor data after validated.

Reviewer: Could you be a little more specific on what is `validated`? It might get confusing for newcomers :stuck_out_tongue:

Reviewer 2: Done.
